optimize sort

VladVega opened the pull request:
Sort optimization with Juttle 7 bump.

fixes https://github.com/juttle/juttle-sql-adapter-common/issues/36

Commit:
optimize sort

Commit:
support juttle 7

VladVega commented:
@davidvgalbraith @bkutil @mstemm @demmer

davidvgalbraith commented on lib/read.js:
```diff
@@ -95,6 +93,20 @@ class ReadSql extends AdapterRead {
+    addSorting() {
+        if (this.sortFields) {
+            // adds multiple sort orders
+            this.sortFields.forEach((sortObj) => {
```
I would put this down by line 128 in `addOptimizations`, seems a little scattered to have it here.

VladVega replied:
Oh. I thought having all the order by statements together would be easiest to read since this function determines sort order. No?

davidvgalbraith replied:
You're the boss! In my head even though they both use the `order by` clause they're pretty unrelated, I'd rather have all the code that deals with `sort` optimization together.

VladVega replied:
I'll think on it.

davidvgalbraith commented on lib/read.js:
```diff
@@ -335,6 +356,13 @@ class ReadSql extends AdapterRead {
                 };
             }
+            
+            if (this.sortFields) {
```
Can SQL not paginate sorted data?

VladVega replied:
You mean using offset? Yes but using offset when data could be coming in in real time could cause errors in the pagination. this method takes real time data into account so IMO it's a superior pagination method.

davidvgalbraith replied:
I guess I don't understand what this is doing...if we read with a limit of `X` points a time range `t1-t2` with `2X` points in it, and optimize a `sort` on that, won't we return only the first `X` points but set the `readEnd` to `t2` and never get the second page?

VladVega replied:
@davidvgalbraith you're absolutely right - the pagination was broken and untested. Fixed now. Take another look?

Commit:
fix sort pagination

VladVega commented:
(also - I'm aware I need to rebase with master on this one - will do so after the +1 so that the latest commit acts as a diff to the previous code that you've already looked at.)

davidvgalbraith commented on lib/read.js:
```diff
@@ -217,6 +238,15 @@ class ReadSql extends AdapterRead {
+        if (this.sortBorderValue) {
+            query = query.where(
+                this.sortFields[0].field,
+                this.sortFields[0].direction === 'desc' ? '<=' : '>=',
```
How are we making sure we don't get duplicate points if there's a bunch of points with `sortBorderValue` as their value spanning two pages?

VladVega replied:
So the way this works is that each result set consists of all values of the border value. For example, lets say the `borderVal`s were `1,2,2,3,3,4,5` and the `fetchSize` is `4` then only `1,2,2` would be returned: the last value becomes the border value none of those will be in the result set. Then the next result set is `>=3`.  In other words, points with the border value are not included in result set. If the entire fetch has the same border value then we throw an error.

davidvgalbraith replied:
I see, and that's in `processPaginatedResults` called with the sort field.

davidvgalbraith commented on test/utils.js:
```diff
@@ -167,8 +164,15 @@ var TestUtils = {
+        
+        if (massageOptions.sort) {
+            massageOptions.sort.forEach(function(sortField) {
+                arr = _.sortBy(sortField);
```
This sets the value of `arr` to a sorted array of the characters in `sortField`...

VladVega replied:
Good catch, the test massaging was making all the testing using the sort massage inaccurate. Fixed now.

Commit:
fix sort testing

1. array must be provided to sortBy method.
2. Take out head proc from test juttles because head cuts off the
result set before the massaging which makes the massaging ineffective

davidvgalbraith commented on lib/optimize.js:
```diff
@@ -106,6 +106,28 @@ class Optimizer {
+        }
+        
+        let groupby = graph.node_get_option(sort, 'groupby');
+        if (groupby) {
```
It's a little more canonical to say `if (graph.node_has_option(sort, 'groupby')) {` in cases like this.

davidvgalbraith commented:
+1

VladVega commented:
still requires a bit of work to clean up. Let me know if someone wants this finished.

davidvgalbraith commented:
What's missing? We've come too far to give up now!
